rcl_yaml_param_parser/src/parser.c

```diff
@@ -0,0 +1,1336 @@
+// Copyright 2018 Apex.AI, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <errno.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <yaml.h>
+
+#include "rcl_yaml_param_parser/parser.h"
+#include "rcl_yaml_param_parser/types.h"
+
+#include "rcl/error_handling.h"
+#include "rcl/types.h"
+#include "rcutils/strdup.h"
+
+/// NOTE: Will allow a max YAML mapping depth of 5
+/// map level 1 : Top level mapping of the config file
+/// map level 2 : Node namespace mapping
+/// map level 3 : Node name mapping
+/// map level 4 : Params mapping
+/// map level 5 : Params namespace mapping
+typedef enum yaml_map_lvl_e
+{
+  MAP_UNINIT_LVL = 0U,
+  MAP_NODE_NAME_LVL = 1U,
+  MAP_PARAMS_LVL = 2U,
+} yaml_map_lvl_t;
+
+/// Basic supported data types in the yaml file
+typedef enum data_types_e
+{
+  DATA_TYPE_UNKNOWN = 0U,
+  DATA_TYPE_BOOL = 1U,
```

Comment:
Will add it when BYTE support is provided. As this is internal to the code and not public, it should be OK. The main reason I did not add is, if we add BYTE type here, then need to provide a empty switch statement for BYTE otherwise MISRA linters will complain